File: 03_use_cases/13_macos_cua/src/metacua/TerminalAttention.swift
import AppKit
import Darwin
import Foundation

final class TerminalAttention {
  private let terminalApp: NSRunningApplication?
  private let enabled: Bool

  init(enabled: Bool = isatty(STDOUT_FILENO) == 1) {
    self.enabled = enabled
    terminalApp = enabled ? TerminalAttention.findLaunchingTerminalApp() : nil
  }

  @discardableResult
  func wake() -> Bool {
    guard enabled else { return false }
    FileHandle.standardOutput.write(Data("\u{7}".utf8))
    guard let terminalApp else { return false }

    let activate = {
      if #available(macOS 14.0, *) {
        _ = terminalApp.activate()
      } else {
        _ = terminalApp.activate(options: [.activateIgnoringOtherApps])
      }
    }

    if Thread.isMainThread {
      activate()
    } else {
      DispatchQueue.main.async(execute: activate)
    }
    return true
  }

  @discardableResult
  func moveAside() -> Bool {
    guard enabled, let terminalApp else { return false }

    let hide = {
      _ = terminalApp.hide()
    }

    if Thread.isMainThread {
      hide()
    } else {
      DispatchQueue.main.async(execute: hide)
    }
    return true
  }

  private static func findLaunchingTerminalApp() -> NSRunningApplication? {
    var pid = getppid()
    var seen: Set<pid_t> = []

    while pid > 1 && !seen.contains(pid) {
      seen.insert(pid)
      if let app = NSRunningApplication(processIdentifier: pid), app.activationPolicy == .regular {
        return app
      }
      guard let parent = parentProcessID(of: pid), parent != pid else { break }
      pid = parent
    }

    return fallbackTerminalApp()
  }

  private static func parentProcessID(of pid: pid_t) -> pid_t? {
    var mib: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, pid]
    var info = kinfo_proc()
    var size = MemoryLayout<kinfo_proc>.stride
    let result = sysctl(&mib, u_int(mib.count), &info, &size, nil, 0)
    guard result == 0, size > 0 else { return nil }
    return info.kp_eproc.e_ppid
  }

  private static func fallbackTerminalApp() -> NSRunningApplication? {
    let candidates = terminalAppNameCandidates()
    return NSWorkspace.shared.runningApplications.first { app in
      guard app.activationPolicy == .regular, let name = app.localizedName else { return false }
      return candidates.contains(name)
    }
  }

  private static func terminalAppNameCandidates() -> Set<String> {
    switch ProcessInfo.processInfo.environment["TERM_PROGRAM"] {
    case "Apple_Terminal":
      return ["Terminal"]
    case "iTerm.app":
      return ["iTerm", "iTerm2"]
    case "vscode":
      return ["Visual Studio Code", "VS Code", "VS Code @ FB"]
    case "WezTerm":
      return ["WezTerm"]
    case "WarpTerminal":
      return ["Warp"]
    default:
      return ["Terminal", "iTerm", "iTerm2", "Visual Studio Code", "VS Code", "VS Code @ FB"]
    }
  }
}

File: 03_use_cases/13_macos_cua/python/metacua/terminal_attention.py
"""Ring the bell and reactivate the launching terminal when a goal ends."""

import os
import subprocess
import sys
from typing import Optional, Set

try:
    from AppKit import NSRunningApplication, NSWorkspace
    from AppKit import NSApplicationActivationPolicyRegular

    _HAS_APPKIT = True
except ImportError:  # pragma: no cover
    _HAS_APPKIT = False


class TerminalAttention:
    def __init__(self, enabled: Optional[bool] = None):
        self._enabled = sys.stdout.isatty() if enabled is None else enabled
        self._terminal_app = self._find_launching_terminal_app() if self._enabled else None

    def wake(self) -> bool:
        if not self._enabled:
            return False
        try:
            sys.stdout.write("\a")
            sys.stdout.flush()
        except (OSError, ValueError):
            pass
        if self._terminal_app is None:
            return False
        try:
            # Newer macOS uses the no-argument activate(); options-based activation
            # remains available as a fallback.
            if hasattr(self._terminal_app, "activate"):
                self._terminal_app.activate()
            else:  # pragma: no cover
                self._terminal_app.activateWithOptions_(1)
        except Exception:  # noqa: BLE001
            return False
        return True

    def _find_launching_terminal_app(self):
        if not _HAS_APPKIT:
            return None
        pid = os.getppid()
        seen: Set[int] = set()
        while pid > 1 and pid not in seen:
            seen.add(pid)
            app = NSRunningApplication.runningApplicationWithProcessIdentifier_(pid)
            if app is not None and app.activationPolicy() == NSApplicationActivationPolicyRegular:
                return app
            parent = self._parent_process_id(pid)
            if parent is None or parent == pid:
                break
            pid = parent
        return self._fallback_terminal_app()

    @staticmethod
    def _parent_process_id(pid: int) -> Optional[int]:
        try:
            out = subprocess.run(
                ["/bin/ps", "-o", "ppid=", "-p", str(pid)],
                capture_output=True,
                text=True,
                timeout=2.0,
            )
        except (OSError, subprocess.SubprocessError):
            return None
        raw = out.stdout.strip()
        try:
            return int(raw)
        except ValueError:
            return None

    def _fallback_terminal_app(self):
        candidates = self._terminal_app_name_candidates()
        for app in NSWorkspace.sharedWorkspace().runningApplications():
            if app.activationPolicy() != NSApplicationActivationPolicyRegular:
                continue
            name = app.localizedName()
            if name in candidates:
                return app
        return None

    @staticmethod
    def _terminal_app_name_candidates() -> Set[str]:
        term = os.environ.get("TERM_PROGRAM")
        if term == "Apple_Terminal":
            return {"Terminal"}
        if term == "iTerm.app":
            return {"iTerm", "iTerm2"}
        if term == "vscode":
            return {"Visual Studio Code", "VS Code"}
        if term == "WezTerm":
            return {"WezTerm"}
        if term == "WarpTerminal":
            return {"Warp"}
        return {"Terminal", "iTerm", "iTerm2", "Visual Studio Code", "VS Code"}
